File: src/main/java/dev/ambryn/task/controllers/TaskController.java
package dev.ambryn.task.controllers;

import dev.ambryn.task.models.Task;
import dev.ambryn.task.services.TaskRunnerService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/tasks")
public class TaskController {
    @Autowired
    TaskRunnerService trs;

    @PostMapping
    public ResponseEntity<Iterable<Task>> createTask(@RequestBody List<Task> tasks) {
        this.trs.addTasks(tasks);
        return ResponseEntity.ok(tasks);
    }

    @GetMapping
    public ResponseEntity<Iterable<Task>> getTasks(@RequestParam("status") Optional<String> oStatus) {
        if (oStatus.isPresent()) {
            String status = oStatus.get();
            switch (status) {
                case "running" -> {
                    return ResponseEntity.ok(this.trs.getRunningTasks());
                }
                case "queued" -> {
                    return ResponseEntity.ok(this.trs.getQueuedTasks());
                }
                case "finished" -> {
                    return ResponseEntity.ok(this.trs.getFinishedTasks());
                }
                case "error" -> {
                    return ResponseEntity.ok(this.trs.getErroredTasks());
                }
                case "cancelled" -> {
                    return ResponseEntity.ok(this.trs.getCancelledTasks());
                }
            }
        }

        return ResponseEntity.ok(this.trs.getTasks());
    }

    @GetMapping("/{id}")
    public ResponseEntity<Task> getTask(@PathVariable Integer id) {
        return ResponseEntity.ok(this.trs.getTask(id));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> cancelTask(@PathVariable Integer id) {
        this.trs.cancelTask(id);
        return ResponseEntity.ok(this.trs.getTask(id));
    }
}

File: src/main/java/dev/ambryn/task/controllers/TaskController.kt
package dev.ambryn.task.controllers

import dev.ambryn.task.models.*
import dev.ambryn.task.services.TaskRunnerService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/tasks")
class TaskController {
    @Autowired
    lateinit var trs: TaskRunnerService

    @PostMapping
    fun createTask(@RequestBody tasks: List<Task>?): ResponseEntity<Iterable<Task>?> {
        trs!!.addTasks(tasks)
        return ResponseEntity.ok(tasks)
    }

    @GetMapping
    fun getTasks(@RequestParam("status") oStatus: Optional<String?>): ResponseEntity<Collection<Task?>?> {
        if (oStatus.isPresent) {
            val status = oStatus.get()
            when (status) {
                "running" -> {
                    return ResponseEntity.ok(trs.runningTasks)
                }

                "queued" -> {
                    return ResponseEntity.ok(trs.queuedTasks)
                }

                "finished" -> {
                    return ResponseEntity.ok(trs.finishedTasks)
                }

                "error" -> {
                    return ResponseEntity.ok(trs.erroredTasks)
                }

                "cancelled" -> {
                    return ResponseEntity.ok(trs.cancelledTasks)
                }
            }
        }
        return ResponseEntity.ok(trs.getTasks())
    }

    @GetMapping("/{id}")
    fun getTask(@PathVariable id: Int): ResponseEntity<Task?> {
        return ResponseEntity.ok(trs.getTask(id))
    }

    @DeleteMapping("/{id}")
    fun cancelTask(@PathVariable id: Int): ResponseEntity<*> {
        trs!!.cancelTask(id)
        return ResponseEntity.ok(trs.getTask(id))
    }
}
